File: TreeMachine.Pro_Cpp/TreeMachine.Pro.Tests/Sources/TreeMachine/Tests_02.cpp
#include "gtest/gtest.h"
#include "Tree.h"

namespace TreeMachine {
    using namespace std;

    TEST(Tests_02, Test_00) { // NOLINT
        auto *const tree = new Tree();
        auto *const root = new Root();
        auto *const a = new A();
        auto *const b = new B();

        for (const auto *const node : vector<NodeBase *>{root, a, b}) {
            EXPECT_NE(node, nullptr);
            EXPECT_EQ(node->Tree(), nullptr);
            EXPECT_EQ(node->TreeRecursive(), nullptr);
            EXPECT_EQ(node->Parent(), nullptr);
            EXPECT_EQ(node->Activity(), NodeBase::EActivity::Inactive);
            EXPECT_EQ(node->Children().size(), 0);
        }

        root->OnBeforeActivateCallback([root, a, b]([[maybe_unused]] const any arg) {
            root->AddChildren(vector<NodeBase *>{a, b}, nullptr);
        });
        root->OnAfterDeactivateCallback([root]([[maybe_unused]] const any arg) {
            root->RemoveChildren(nullptr, [](auto *const child, [[maybe_unused]] const auto arg) { delete child; });
        });

        {
            // AddRoot
            tree->AddRoot(root, nullptr);
            EXPECT_NE(tree->Root(), nullptr);
            EXPECT_EQ(tree->Root()->Tree(), tree);
            EXPECT_EQ(tree->Root()->TreeRecursive(), tree);
            EXPECT_EQ(tree->Root()->Parent(), nullptr);
            EXPECT_EQ(tree->Root()->Activity(), NodeBase::EActivity::Active);
            EXPECT_EQ(tree->Root()->Children().size(), 2);
            for (const auto *const child : tree->Root()->Children()) {
                EXPECT_EQ(child->Tree(), nullptr);
                EXPECT_EQ(child->TreeRecursive(), tree);
                EXPECT_EQ(child->Parent(), tree->Root());
                EXPECT_EQ(child->Activity(), NodeBase::EActivity::Active);
                EXPECT_EQ(child->Children().size(), 0);
            }
        }
        {
            // RemoveRoot
            tree->RemoveRoot(nullptr, [](auto *const root, [[maybe_unused]] const auto arg) { delete root; }); // NOLINT
            EXPECT_EQ(tree->Root(), nullptr);
        }

        delete tree;
    }

    TEST(Tests_02, Test_01) { // NOLINT
        auto *const tree = new Tree();
        auto *const root = new Root();
        auto *const a = new A();
        auto *const b = new B();

        for (const auto *const node : vector<NodeBase *>{root, a, b}) {
            EXPECT_NE(node, nullptr);
            EXPECT_EQ(node->Tree(), nullptr);
            EXPECT_EQ(node->TreeRecursive(), nullptr);
            EXPECT_EQ(node->Parent(), nullptr);
            EXPECT_EQ(node->Activity(), NodeBase::EActivity::Inactive);
            EXPECT_EQ(node->Children().size(), 0);
        }

        root->OnAfterActivateCallback([root, a, b]([[maybe_unused]] const any arg) {
            root->AddChildren(vector<NodeBase *>{a, b}, nullptr);
        });
        root->OnBeforeDeactivateCallback([root]([[maybe_unused]] const any arg) {
            root->RemoveChildren(nullptr, [](auto *const child, [[maybe_unused]] const auto arg) { delete child; });
        });

        {
            // AddRoot
            tree->AddRoot(root, nullptr);
            EXPECT_NE(tree->Root(), nullptr);
            EXPECT_EQ(tree->Root()->Tree(), tree);
            EXPECT_EQ(tree->Root()->TreeRecursive(), tree);
            EXPECT_EQ(tree->Root()->Parent(), nullptr);
            EXPECT_EQ(tree->Root()->Activity(), NodeBase::EActivity::Active);
            EXPECT_EQ(tree->Root()->Children().size(), 2);
            for (const auto *const child : tree->Root()->Children()) {
                EXPECT_EQ(child->Tree(), nullptr);
                EXPECT_EQ(child->TreeRecursive(), tree);
                EXPECT_EQ(child->Parent(), tree->Root());
                EXPECT_EQ(child->Activity(), NodeBase::EActivity::Active);
                EXPECT_EQ(child->Children().size(), 0);
            }
        }
        {
            // RemoveRoot
            tree->RemoveRoot(nullptr, [](auto *const root, [[maybe_unused]] const auto arg) { delete root; }); // NOLINT
            EXPECT_EQ(tree->Root(), nullptr);
        }

        delete tree;
    }

}

File: TreeMachine.Pro_CSharp/TreeMachine.Pro.Tests/System.TreeMachine/Tests_02.cs
namespace System.TreeMachine {
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;
    using NUnit.Framework;
    using Assert = NUnit.Framework.Assert;

    public class Tests_02 {

        [Test]
        public void Test_00() {
            var tree = new Tree();
            var root = new Root();
            var a = new A();
            var b = new B();

            root.OnBeforeActivateCallback += arg => {
                root.AddChildren( [ a, b ], null );
            };
            root.OnAfterDeactivateCallback += arg => {
                root.RemoveChildren( null, null );
            };

            {
                // tree.AddRoot root
                tree.AddRoot( root, null );
                Assert.That( tree.Root, Is.EqualTo( root ) );
                Assert.That( tree.Root.Tree, Is.EqualTo( tree ) );
                Assert.That( tree.Root.Tree_NoRecursive, Is.EqualTo( tree ) );
                Assert.That( tree.Root.IsRoot, Is.True );
                Assert.That( tree.Root.Root, Is.EqualTo( root ) );
                Assert.That( tree.Root.Parent, Is.Null );
                Assert.That( tree.Root.Ancestors.Count(), Is.EqualTo( 0 ) );
                Assert.That( tree.Root.AncestorsAndSelf.Count(), Is.EqualTo( 1 ) );
                Assert.That( tree.Root.Activity, Is.EqualTo( Node.Activity_.Active ) );
                Assert.That( tree.Root.Children.Count, Is.EqualTo( 2 ) );
                Assert.That( tree.Root.Descendants.Count, Is.EqualTo( 2 ) );
                Assert.That( tree.Root.DescendantsAndSelf.Count, Is.EqualTo( 3 ) );
                foreach (var child in tree.Root.Children) {
                    Assert.That( child, Is.EqualTo( a ).Or.EqualTo( b ) );
                    Assert.That( child.Tree, Is.EqualTo( tree ) );
                    Assert.That( child.Tree_NoRecursive, Is.Null );
                    Assert.That( child.IsRoot, Is.False );
                    Assert.That( child.Root, Is.EqualTo( root ) );
                    Assert.That( child.Parent, Is.EqualTo( root ) );
                    Assert.That( child.Ancestors.Count(), Is.EqualTo( 1 ) );
                    Assert.That( child.AncestorsAndSelf.Count(), Is.EqualTo( 2 ) );
                    Assert.That( child.Activity, Is.EqualTo( Node.Activity_.Active ) );
                    Assert.That( child.Children.Count, Is.EqualTo( 0 ) );
                    Assert.That( child.Descendants.Count, Is.EqualTo( 0 ) );
                    Assert.That( child.DescendantsAndSelf.Count, Is.EqualTo( 1 ) );
                }
            }
            {
                // tree.RemoveRoot
                tree.RemoveRoot( null, null );
                Assert.That( tree.Root, Is.Null );
            }
        }

        [Test]
        public void Test_02() {
            var tree = new Tree();
            var root = new Root();
            var a = new A();
            var b = new B();

            root.OnAfterActivateCallback += arg => {
                root.AddChildren( [ a, b ], null );
            };
            root.OnBeforeDeactivateCallback += arg => {
                root.RemoveChildren( null, null );
            };

            {
                // tree.AddRoot root
                tree.AddRoot( root, null );
                Assert.That( tree.Root, Is.EqualTo( root ) );
                Assert.That( tree.Root.Tree, Is.EqualTo( tree ) );
                Assert.That( tree.Root.Tree_NoRecursive, Is.EqualTo( tree ) );
                Assert.That( tree.Root.IsRoot, Is.True );
                Assert.That( tree.Root.Root, Is.EqualTo( root ) );
                Assert.That( tree.Root.Parent, Is.Null );
                Assert.That( tree.Root.Ancestors.Count(), Is.EqualTo( 0 ) );
                Assert.That( tree.Root.AncestorsAndSelf.Count(), Is.EqualTo( 1 ) );
                Assert.That( tree.Root.Activity, Is.EqualTo( Node.Activity_.Active ) );
                Assert.That( tree.Root.Children.Count, Is.EqualTo( 2 ) );
                Assert.That( tree.Root.Descendants.Count, Is.EqualTo( 2 ) );
                Assert.That( tree.Root.DescendantsAndSelf.Count, Is.EqualTo( 3 ) );
                foreach (var child in tree.Root.Children) {
                    Assert.That( child, Is.EqualTo( a ).Or.EqualTo( b ) );
                    Assert.That( child.Tree, Is.EqualTo( tree ) );
                    Assert.That( child.Tree_NoRecursive, Is.Null );
                    Assert.That( child.IsRoot, Is.False );
                    Assert.That( child.Root, Is.EqualTo( root ) );
                    Assert.That( child.Parent, Is.EqualTo( root ) );
                    Assert.That( child.Ancestors.Count(), Is.EqualTo( 1 ) );
                    Assert.That( child.AncestorsAndSelf.Count(), Is.EqualTo( 2 ) );
                    Assert.That( child.Activity, Is.EqualTo( Node.Activity_.Active ) );
                    Assert.That( child.Children.Count, Is.EqualTo( 0 ) );
                    Assert.That( child.Descendants.Count, Is.EqualTo( 0 ) );
                    Assert.That( child.DescendantsAndSelf.Count, Is.EqualTo( 1 ) );
                }
            }
            {
                // tree.RemoveRoot
                tree.RemoveRoot( null, null );
                Assert.That( tree.Root, Is.Null );
            }
        }

    }
}
